File: java/dtdgen-master/src/main/java/com/philhanna/dtdgen/ElementModel.java
package com.philhanna.dtdgen;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

import com.philhanna.dtdgen.dtd.DTDGenerator;

/**
 * Keeps track of the the possible contents of an element, based on all
 * instances of it found in the source XML.
 */
public class ElementModel {

   // ==================================================================
   // Class constants and variables
   // ==================================================================

   // ==================================================================
   // Class methods
   // ==================================================================

   // ==================================================================
   // Instance variables
   // ==================================================================

   private final String name;
   private int occurrences = 0;
   private boolean hasCharacterContent = false;
   private boolean sequenced = true;
   private final List<ChildModel> childseq = new ArrayList<ChildModel>();
   private final Map<String, AttributeModel> attributes = new LinkedHashMap<String, AttributeModel>();

   /**
    * Minimum number of attribute values that must appear for the
    * attribute to be regarded as an ID value
    */
   protected static final int MIN_ID_VALUES = 10;

   // ==================================================================
   // Constructors
   // ==================================================================

   public ElementModel(String name) {
      this.name = name;
   }

   // ==================================================================
   // Implementation of ElementModel interface
   // ==================================================================

   /**
    * Returns the element tag name
    */
   public String getName() {
      return name;
   }

   /**
    * Returns the number of times this element was found in the input
    * XML
    */
   public int getOccurrences() {
      return occurrences;
   }

   /**
    * Returns the <code>ChildModel</code> for the child element with the
    * specified name
    * @param name the name of the child element
    * @return the <code>ChildModel</code> element, or <code>null</code>,
    *         if this element does not have a child with that name.
    */
   public ChildModel getChildModel(String name) {
      for (final ChildModel childDetails : childseq) {
         if (childDetails.getName().equals(name))
            return childDetails;
      }
      return null;
   }

   /**
    * Returns the number of <code>ChildModel</code> elements this parent
    * element has
    */
   public int getChildModelCount() {
      return childseq.size();
   }

   /**
    * Returns the specified <code>ChildModel</code> element
    * @param index the index within the list of children (zero-based)
    */
   public ChildModel getChildModel(int index) {
      return childseq.get(index);
   }

   /**
    * Returns <code>true</code> if this element can have character
    * content
    */
   public boolean hasCharacterContent() {
      return hasCharacterContent;
   }

   /**
    * Returns <code>true</code> if this element's children always occur
    * in a specific sequence
    */
   public boolean isSequenced() {
      return sequenced;
   }

   /**
    * Returns an iterator over the list of attribute names (in the order
    * that they first appear in the DTD)
    */
   public Iterator<String> attributeNameIterator() {
      return attributes.keySet().iterator();
   }

   /**
    * Returns the attribute with the specified name, or
    * <code>null</code>, if no such attribute exists
    * @param name the attribute name
    */
   public AttributeModel getAttributeModel(String name) {
      return attributes.get(name);
   }

   /**
    * Returns the name of the ID attribute, if one can be inferred,
    * otherwise returns <code>null</code>.
    */
   public String getIDAttributeName() {

      for (final String attrName : attributes.keySet()) {
         final AttributeModel attributeModel = attributes.get(attrName);

         /*
          * If every value of the attribute is distinct, and there are
          * at least MIN_ID_VALUES, treat it as an ID. ID values must be
          * Names. Only allowed one ID attribute per element type.
          */

         if (attributeModel.isAllNames()
               && (attributeModel.isUnique())
               && (attributeModel
                     .getOccurrences() >= ElementModel.MIN_ID_VALUES))
            return attrName;

         /*
          * TODO: This may give the wrong answer. We should check
          * whether the value sets of two candidate-ID attributes
          * overlap, in which case they can't both be IDs !!
          */

      }
      return null;
   }

   public void setHasCharacterContent(boolean hasCharacterContent) {
      this.hasCharacterContent = hasCharacterContent;
   }

   public void setSequenced(boolean sequenced) {
      this.sequenced = sequenced;
   }

   public void incrementOccurrences() {
      occurrences++;
   }

   public AttributeModel getAttribute(String attributeName) {
      return attributes.get(attributeName);
   }

   public void addAttribute(AttributeModel attributeModel) {
      attributes.put(attributeModel.getName(), attributeModel);
   }

   public String[] getATTLISTs() {
      return DTDGenerator.getATTLISTs(this);
   }

   public void addChild(ChildModel childModel) {
      childseq.add(childModel);
   }
}

File: dtdgen/element_model.py
from typing import List

from dtdgen import ChildModel, AttributeModel


class ElementModel:
    """Keeps track of the possible contents of an element, based on all
    instances of it found in the source XML."""

    # Minimum number of attribute values that must appear for the
    # attribute to be regarded as an ID value
    MIN_ID_VALUES = 10

    def __init__(self, name: str):
        """ Creates a new ElementModel with the specified name """
        self._name: str = name
        self._occurrences: int = 0
        self._character_content: bool = False
        self._sequenced: bool = True
        self._childseq: List[ChildModel] = []
        self._attributes: dict[str, AttributeModel] = dict()

    @property
    def name(self):
        return self._name

    @property
    def occurrences(self):
        """ Returns the number of times this element was found
            in the input XML """
        return self._occurrences

    @occurrences.setter
    def occurrences(self, value: int):
        self._occurrences = value

    def increment_occurrences(self):
        self._occurrences += 1

    @property
    def has_character_content(self):
        return self._character_content

    @has_character_content.setter
    def has_character_content(self, value: bool):
        self._character_content = value

    @property
    def is_sequenced(self):
        return self._sequenced

    @is_sequenced.setter
    def is_sequenced(self, value: bool):
        self._sequenced = value

    def child_iterator(self):
        for child in self._childseq:
            yield child

    def get_child_model(self, index: int = None, name: str = None):
        """Returns the child model by name or index, or None if it does not exist."""
        if index is not None:
            return self._childseq[index]
        if name is not None:
            child = [x for x in self._childseq if x.name == name][0]
            return child

    def get_child_model_count(self):
        """Returns the number of ChildModel elements this parent has."""
        return len(self._childseq)

    def add_child(self, child_model: ChildModel):
        """ Adds a child element for this element """
        self._childseq.append(child_model)

    def attribute_names(self):
        """ Generator that yields attribute names, one for each call """
        for attrname in self._attributes:
            yield attrname

    def get_attribute_model(self, name: str) -> AttributeModel:
        return self._attributes.get(name, None)

    def add_attribute(self, attribute_model: AttributeModel) -> None:
        attrname: str = attribute_model.name
        self._attributes[attrname] = attribute_model

    def id_attribute_name(self):
        for attr_name, attr_model in self._attributes.items():
            # If every value of the attribute is distinct, and there are
            # at least MIN_ID_VALUES, treat it as an ID. ID values must be
            # Names. Only allowed one ID attribute per element type.
            if all([
                attr_model.all_names,
                attr_model.unique,
                attr_model.occurrences >= ElementModel.MIN_ID_VALUES,
            ]):
                return attr_name

            # TODO: This may give the wrong answer. We should check
            # whether the value sets of two candidate-ID attributes
            # overlap, in which case they can't both be IDs !!

        return None



